INT B-23350 - story 2 enter edit gun safe sc

taeJungCaci opened the pull request:
## [B-23350](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-23350)

## Summary

This is the User Story 2 for the gun safe feature and 1st one can be found [here](https://github.com/transcom/mymove/pull/15458). Basically adding gun safe option to PPM shipments and view Allowances page similar to how PPM pro-gears are done. This covers the flow for the office user editing/creating PPM shipments and View allowances page.


### How to test

go to your `.envrc` and set `export FEATURE_FLAG_GUN_SAFE=true` and `export DTOD_USE_MOCK=true`

1. Create a PPM move as a customer.
2. As a SC -> Edit shipment and scroll down and you should see options for Gun safe with `Yes` or `No`
3. Choose `Yes` and verify that the weight field becomes visible and you are not able to enter more than 500 lbs.
4. Verify the note message below the field to be:
**"The government authorizes the shipment of a gun safe up to 500 lbs. The weight entitlement is charged for any weight over 500 lbs. The gun safe weight cannot be added to overall entitlement for O-6 and higher ranks."**
5. Save and on the PPM shipment card, click the expand arrow icon to see gun safe field with the entered value.
6. Click "Edit Allowances" and verify that Gun safe weight field is visible and that you are not able to enter more than 500 lbs.
7. Save and check the Allowances card for gun safe weight field.
8. Test the same for TOO and creating PPMs.


## Screenshots
![image](https://github.com/user-attachments/assets/e8d5c550-96fb-4508-b668-72a3f734ae3f)
![image](https://github.com/user-attachments/assets/4e550783-76e7-4f95-873a-2984bd83b8cb)
![image](https://github.com/user-attachments/assets/90b222f2-5cf0-47b7-b874-0eef8c672946)
![image](https://github.com/user-attachments/assets/19d4792d-0b45-4653-84b5-ddbb97f67a74)

dritter44 reviewed (dismissed):
Code changes are within standard practice. No ts comments left. Tests pass locally. Approved.

KonstanceH reviewed (approved):
Functionality working as intended. Tests are passing locally and code looks good to me. LGTM!

danieljordan-caci commented on src/components/Office/AllowancesDetailForm/AllowancesDetailForm.jsx:
```diff
@@ -36,6 +38,7 @@ const AllowancesDetailForm = ({ header, entitlements, branchOptions, formIsDisab
       if (enabled) {
         setEnableUB(true);
       }
+      setIsGunSafeEnabled(await isBooleanFlagEnabled(FEATURE_FLAG_KEYS.GUN_SAFE));
```
Can you update the func name here since it's no longer _just_ checking the UB feature flag?

taeJungCaci replied:
Good catch - updated

taeJungCaci commented on pkg/services/ppmshipment/ppm_shipment_creator.go:
```diff
@@ -199,6 +199,34 @@ func (f *ppmShipmentCreator) createPPMShipment(appCtx appcontext.AppContext, ppm
+				return err
+			}
+
+			entitlement := move.Orders.Entitlement
```
Good catch! Added a nil check and tests. Also had the same thing done for `ppm_shipment_updater.go`

r-mettler commented on pkg/handlers/ghcapi/orders_test.go:
```diff
@@ -1720,10 +1809,66 @@ func (suite *HandlerSuite) TestCounselingUpdateAllowanceHandler() {
 		suite.Equal(*body.StorageInTransit, *ordersPayload.Entitlement.StorageInTransit)
 	})
 
+	suite.Run("Returns 200 when all validations pass - gun safe FF off", func() {
```
Nit: Test on line 1518 above has the same name, can you make one of these tests a bit more descriptive to help distinguish between the two if the test ever fails? 
Maybe something like ```suite.Run("Returns 200 when all validations pass - gun safe FF off, SC only", func() {```

danieljordan-caci reviewed (changes requested):
Few things here after going through the flow.

First thing might not be related, but I'd like to request it anyway.

The `Required` label on the customer side is driving me nuts. Should be on the right like the rest of the ones above
![Screenshot 2025-05-29 at 9 21 19 AM](https://github.com/user-attachments/assets/438ceef7-5f26-42f1-addf-6d0a591c30c8)


All looks good on the SC side, seeing the new label for gun safe
![Screenshot 2025-05-29 at 9 22 35 AM](https://github.com/user-attachments/assets/dd91609e-2692-48a3-a1bb-b4529ccf29b4)

I would like to make a request to avoid having vertically aligned radio groups when there's more than one. So this:
![Screenshot 2025-05-29 at 9 22 48 AM](https://github.com/user-attachments/assets/662a8e1b-46b8-475a-856a-13eb25a2be66)

could be changed to this, which is easier to look at imo:
![Screenshot 2025-05-29 at 9 45 03 AM](https://github.com/user-attachments/assets/ef7957f5-6374-412d-8414-740fa0905f78)

This was done by adding some styling
```
            <Label className={styles.radioLabel}>Pro gear?</Label>
            <FormGroup className={styles.radioGroup}>
```
and
```
                <Label className={styles.radioLabel}>Gun safe?</Label>
                <FormGroup className={styles.radioGroup}>
```
and then the CSS
```
.radioLabel {
  margin-bottom: 0 !important;
}

.radioGroup {
  display: flex;
  gap: 15px;
  margin-top: 0 !important;
}
```

Once that was done, it looks really nice!
![Screenshot 2025-05-29 at 9 46 03 AM](https://github.com/user-attachments/assets/7281b056-f520-4f2b-abf7-f137fe9fcde0)

Last thing... on the move history, I think we should have the label say `Updated allowances` instead of `Update shipment` (looking at the third one down here). The shipment is being updated, but the allowance is.
![Screenshot 2025-05-29 at 9 46 44 AM](https://github.com/user-attachments/assets/2e1c67c7-a3ee-4081-a243-808185a13a25)

r-mettler reviewed (changes requested):
Need playwright tests covering selection of gun safe. In particular look at tests in ```playwright/tests/office/ppms/happyPathFlows.spec.js``` This handles PPM related moves from Office side. ```test('is able to edit a PPM shipment', async ({ page, ppmPage })``` is currently failing on the Government constructed cost not matching but not sure if this is related to the changes made by this backlog or not.

taeJungCaci commented on src/constants/MoveHistory/Database/WeightFields.js:
```diff
@@ -4,6 +4,8 @@ export default {
   authorized_weight: 'authorized_weight',
   pro_gear_weight: 'pro_gear_weight',
   pro_gear_weight_spouse: 'pro_gear_weight_spouse',
+  gun_safe_weight: 'pro_gear_weight',
```
Thanks for catching that-  Fixed.

taeJungCaci commented:
> Need playwright tests covering selection of gun safe. In particular look at tests in `playwright/tests/office/ppms/happyPathFlows.spec.js` This handles PPM related moves from Office side. `test('is able to edit a PPM shipment', async ({ page, ppmPage })` is currently failing on the Government constructed cost not matching but not sure if this is related to the changes made by this backlog or not.
> 
> Also take a look at adding a test that includes selecting Gun Safe `Yes` in the `playwright/tests/office/ppms/addShipment.spec.js` file.

Added some playwright tests and fixed the ones in that directory. It was broken due to some dates not matching that was updated in one of the build functions. Also I checked the latest main branch that does not have my changes and it still broke for the Government constructed cost not matching - probably some pricing changed in the db I bet. So I just removed the dollar amount in the check. Let me know if these changes are good enough.

danieljordan-caci reviewed (dismissed):
reapproving

taeJungCaci commented:
> reapproving

Sorry, requested to wrong Daniel 😆

dritter44 reviewed (approved):
Just noticed two areas where there is commented out code.

taeJungCaci commented on pkg/services/ppmshipment/ppm_shipment_creator.go:
```diff
@@ -206,6 +206,38 @@ func (f *ppmShipmentCreator) createPPMShipment(appCtx appcontext.AppContext, ppm
+			}
+
+			entitlement.GunSafe = *ppmShipment.HasGunSafe
+			// maxGunSafeWeight := 0
```
Good catch. I was supposed to delete those. Deleted - thanks!

taeJungCaci commented on pkg/services/ppmshipment/ppm_shipment_updater.go:
```diff
@@ -319,20 +321,24 @@ func (f *ppmShipmentUpdater) updatePPMShipment(appCtx appcontext.AppContext, ppm
-					}
-				}
-				entitlement.GunSafeWeight = maxGunSafeWeight
+				// maxGunSafeWeight := 0
```
Deleted

r-mettler commented on playwright/tests/office/ppms/happyPathFlows.spec.js:
```diff
@@ -25,7 +27,7 @@ test.describe('Services counselor user', () => {
 
     // Verify SIT info
-    await expect(page.getByText('Government constructed cost: $326')).toBeVisible();
+    await expect(page.getByText('Government constructed cost:')).toBeVisible();
```
👍 Love this. Thanks. We have so many hardcoded values that when something minor changes the tests fail.

r-mettler commented on playwright/tests/office/ppms/viewShipment.spec.js:
```diff
@@ -17,7 +17,7 @@ test.describe('Services counselor user', () => {
     await page.locator('[data-testid="ShipmentContainer"] ').click();
 
-    await expect(page.locator('[data-testid="expectedDepartureDate"]')).toContainText('15 Mar 2020');
+    await expect(page.locator('[data-testid="expectedDepartureDate"]')).toContainText('15 Mar 2024');
```
MVP 🏅 for fixing these dates as well.

danieljordan-caci reviewed (approved):
this feels all too familiar...

dritter44 reviewed (approved):
Functions as described, code changes are within standard practice. Approved!
